gh-75459: Doc: C API: Improve object life cycle documentation (#125962)
python/cpython: Doc/c-api/lifecycle.rst

.. highlight:: c

.. _life-cycle:

Object Life Cycle
=================

Life Events
-----------

The figure below illustrates the order of events that can occur throughout an
object's life [1]_.  An arrow from *A* to *B* indicates that event *B* can
occur after event *A* has occurred, with the arrow's label indicating the
condition that must be true for *B* to occur after *A*.

.. raw:: html

>    :file: lifecycle.dot.svg

Reviewer: Here's how it looks in dark mode:

<details>
<summary>Details</summary>

![image](https://github.com/user-attachments/assets/1af5d60c-9484-40d7-aab6-f38861284147)

</details>

Is it possible to make the SVG transparent and then apply `:class: invert-in-dark-mode`?

Compare https://docs.python.org/3/library/pathlib.html which is `.. image::` rather than `.. raw:: html`:

https://github.com/python/cpython/blob/75f7cf91ec5afc6091a0fd442a1f0435c19300b2/Doc/library/pathlib.rst#L21-L29

We should also have an alt tag.

Author: I got dark mode to work.  I didn't use `.. image::` because that breaks the clickable parts of the image.  SVG doesn't have an alt attribute, but I think I did the accessible equivalent.  I'm not sure how to test it though—any hints would be appreciated.

:c:member:`~PyTypeObject.tp_finalize` and :c:member:`~PyTypeObject.tp_clear` as
needed to destroy a cyclic isolate; this may change in a future version.

Functions
---------

To allocate and free memory, see :ref:`allocating-objects`.

.. c:function:: void PyObject_CallFinalizer(PyObject *op)

   Finalizes the object as described in :c:member:`~PyTypeObject.tp_finalize`.

.. c:function:: int PyObject_CallFinalizerFromDealloc(PyObject *op)

   Same as :c:func:`PyObject_CallFinalizer` but meant to be called at the
   beginning of the object's destructor (:c:member:`~PyTypeObject.tp_dealloc`).
   The object's reference count must already be 0.  If the object's finalizer
   resurrects the object, this function returns -1; no further destruction
   should happen.  Otherwise, this function returns 0 and destruction can
   continue normally.

.. rubric:: Footnotes

.. [1] Disclaimer: Some of the details documented in this section may change in
   a future version of :term:`CPython`.  Care was taken to limit the promises
   made here without sacrificing information needed to correctly design types
   and without sacrificing the ability to improve CPython in a

>    backwards-compatible way.  Time will tell whether we were successful.

Reviewer 2: If this PR is accepted, we won't need this disclaimer, as breaking documented behaviour would be a backwards-incompatible change.

     * :c:member:`~PyTypeObject.tp_new` is called to create a new object.
     * :c:member:`~PyTypeObject.tp_alloc` is directly called by
       :c:member:`~PyTypeObject.tp_new` to allocate the memory for the new
       object.
     * :c:member:`~PyTypeObject.tp_init` initializes the newly created object.
       :c:member:`!tp_init` can be called again to re-initialize an object, if
       desired.

   * After :c:member:`!tp_init` completes, the object is ready to use.
   * After the last reference to an object is removed:

     #. If an object is not marked as *finalized*, it might be finalized by
        marking it as *finalized* and calling its
        :c:member:`~PyTypeObject.tp_finalize` function.  :term:`CPython`
        currently does not finalize an object when the last reference to it is
        deleted, but this may change in the future.
     #. If the object is marked as finalized,
        :c:member:`~PyTypeObject.tp_clear` might be called to clear references
        held by the object.  :term:`CPython` currently does not clear an object
        in response to the deletion of the last reference, but this may change
        in the future.

>      #. :c:member:`~PyTypeObject.tp_dealloc` is called to destroy the object.

Reviewer 3: You should mention that `tp_dealloc` should call `tp_free` here.

>    * The :c:member:`~PyTypeObject.tp_finalize` function can optionally add a

Reviewer 3: I wouldn't call it "optional," because the user doesn't really control whether it gets resurrected or not.
```suggestion
   * The :c:member:`~PyTypeObject.tp_finalize` function can, in some cases, add a
```

Author: It is optional in the sense that users can control whether the object is resurrected or not.

Reviewer 3: Yeah, *users* control it, not necessarily the type.

Author: I don't follow.  The `tp_finalize` function is what adds the reference (directly or indirectly by calling something else), so the object's type is the only thing that has direct control over whether the object is resurrected or not.

Reviewer 3: Typically resurrection isn't intentional by the designer of the type object; it's a side effect of calling some other cleanup code which happens to store a reference.

     reference to the object, *resurrecting* it and preventing its pending
     destruction.  (Only :c:member:`!tp_finalize` can resurrect an object;
     :c:member:`~PyTypeObject.tp_clear` and
     :c:member:`~PyTypeObject.tp_dealloc` cannot.)  Resurrecting an object may
     or may not cause the object's *finalized* mark to be removed.  Currently,
     :term:`CPython` does not remove the *finalized* mark from a resurrected
     object if the object's type supports garbage collection (i.e., the

>      :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set) but does remove the mark if the

Reviewer 3: I don't think there should be distinction between the type and object here. It's sort of implied that if the object supports garbage collection, then its type has marked it as such.
```suggestion
     object if it supports garbage collection, but does remove the mark if the
```

Author: I left the reference because I like for readers to be able to click for more info.

     object does not support garbage collection; either or both of these
     behaviors may change in the future.
   * :c:member:`~PyTypeObject.tp_dealloc` can optionally call
     :c:member:`~PyTypeObject.tp_finalize` via
     :c:func:`PyObject_CallFinalizerFromDealloc` if it wishes to reuse that
     code to help with object destruction.  This is recommended because it
     guarantees that :c:member:`!tp_finalize` is always called before
     destruction.
   * When :c:member:`~PyTypeObject.tp_dealloc` finishes object destruction, it

>      directly calls :c:member:`~PyTypeObject.tp_free` to deallocate the memory.

Reviewer 3: This is possibly a good place to mention `PyObject_GC_Del` or `PyObject_Free` being automatically set in the `tp_free` slot.

   * If the object is a member of a :term:`cyclic isolate` and
     :c:member:`~PyTypeObject.tp_clear` fails to break the reference cycle (or
     that function is not called), the objects remain indefinitely
     uncollectable (they "leak").  See :data:`gc.garbage`.

   If the object is marked as supporting garbage collection (the
   :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
   :c:member:`~PyTypeObject.tp_flags`), the following events are also possible:

>    * The garbage collector occasionally calls

Reviewer 3: Nitpick: I think "periodically" is a better term here.
```suggestion
   * The garbage collector periodically calls
```

Author: I intentionally avoided "periodically" here because I don't want the reader to assume dependable regularity.  Generational GC means that sometimes an object is scanned and sometimes not, and users don't have much control over frequency or regularity.

     :c:member:`~PyTypeObject.tp_traverse` to identify :term:`cyclic isolates
     <cyclic isolate>`.
   * When the garbage collector discovers a :term:`cyclic isolate`, it
     finalizes one of the objects in the group by calling its

>      :c:member:`~PyTypeObject.tp_finalize` function.  This repeats until the

Reviewer 3: ```suggestion
     :c:member:`~PyTypeObject.tp_finalize` function, if it has one.  This repeats until the
```

     cyclic isolate doesn't exist or all of the objects have been finalized.
   * :c:member:`~PyTypeObject.tp_finalize` can resurrect the object by adding a
     reference from outside the :term:`cyclic isolate`.  The new reference
     causes the group of objects to no longer form a cyclic isolate (the

>      reference cycle may still exist; if it does the objects are no longer

Reviewer 3: There's some cause and effect here:
```suggestion
     reference cycle may still exist; but if it does the objects are no longer
```

     isolated).
   * When the garbage collector discovers a :term:`cyclic isolate` and all of
     the objects in the group have already been finalized, the garbage
     collector clears one or more of the uncleared objects in the group
     (possibly concurrently, but with the :term:`GIL` held :ref:`unless

>      disabled <whatsnew313-free-threaded-cpython>`) by calling each's

Reviewer 3: Again, "GIL" is a weird term in the C API, and I don't want people to think that they'll have to call `PyGILState_Ensure` if `Py_GIL_DISABLED` is set. If the GIL is disabled, you don't really know or care as a user, because the core does most of the work for you.
```suggestion
     (possibly concurrently, but in a thread-safe manner) by calling each's
```

     :c:member:`~PyTypeObject.tp_clear` function.  This repeats as long as the
     cyclic isolate still exists and not all of the objects have been cleared.

Cyclic Isolate Destruction
--------------------------

Listed below are the stages of life of a hypothetical :term:`cyclic isolate`
that continues to exist after each member object is finalized or cleared.  It
is a bug if a cyclic isolate progresses through all of these stages; it should
vanish once all objects are cleared, if not sooner.  A cyclic isolate can
vanish either because the reference cycle is broken or because the objects are
no longer isolated due to finalizer resurrection (see
:c:member:`~PyTypeObject.tp_finalize`).

> 0. **Reachable** (pre-cyclic isolate): All objects are in their normal,

Reviewer 3: Nitpick: the dash here makes it look like "'before cyclic' isolate", instead of "'before' cyclic isolate".
```suggestion
0. **Reachable** (pre cyclic-isolate): All objects are in their normal,
```

>    reachable state.  A reference cycle exists, but an external reference means

Reviewer 3: ```suggestion
   reachable state.  A reference cycle could exist, but an external reference means
```

   the objects are not yet isolated.
#. **Unreachable but consistent:** The final reference from outside the group
   of objects has been removed, causing the objects to become isolated (thus a
   cyclic isolate is born).  None of the group's objects have been finalized or

>    cleared yet.

Reviewer 3: Maybe mention that it will remain in this state until the garbage collector runs.

#. **Mix of finalized and not finalized:** Objects in a cyclic isolate are
   finalized one at a time, which means that there is a period of time when the
   cyclic isolate is composed of a mix of finalized and non-finalized objects.
   Finalization order is unspecified, so it can appear random.  A finalized
   object must behave in a sane manner when non-finalized objects interact with
   it, and a non-finalized object must be able to tolerate the finalization of
   an arbitrary subset of its referents.
#. **All finalized:** All objects in a cyclic isolate are finalized before any
   of them are cleared.
#. **Mix of finalized and cleared:** The objects can be cleared serially or

>    concurrently (but with the :term:`GIL` held :ref:`unless disabled
>    <whatsnew313-free-threaded-cpython>`); either way, some will finish before

Reviewer 3: Again, saying "GIL might not be held" is probably misleading. But anyways, user won't really notice or care if they're getting cleared concurrently or not.
```suggestion
   concurrently; either way, some will finish before
```

Author: They will care if they do any inter-object calls that might require locking so I left the GIL comment there, but removed the reference to free threading.

   others.  A finalized object must be able to tolerate the clearing of a
   subset of its referents.  :pep:`442` calls this stage "cyclic trash".
#. **Leaked:** If a cyclic isolate still exists after all objects in the group
   have been finalized and cleared, then the objects remain indefinitely
   uncollectable (see :data:`gc.garbage`).  It is a bug if a cyclic isolate
   reaches this stage---it means the :c:member:`~PyTypeObject.tp_clear` methods
   of the participating objects have failed to break the reference cycle as
   required.

If :c:member:`~PyTypeObject.tp_clear` did not exist, then Python would have no
way to break a reference cycle.  The destruction of an object in a cyclic
isolate would result in a dangling pointer, triggering undefined behavior when
an object referencing the destroyed object is itself destroyed.

:c:member:`~PyTypeObject.tp_finalize` is not needed to safely destroy a cyclic
isolate, but its existence makes it easier to design types that behave in a

> sane manner when objects are cleared.  Clearing an object might necessarily

Reviewer 3: It could also be worth mentioning that it's completely safe to execute arbitary Python code in `tp_finalize`. In a lot of cases, it's not necessarily used for destroying the object, but just for doing something upon deletion (e.g. closing a file).

* A cleared object might have references to or from finalized and cleared
  objects, but not to or from non-finalized objects.

Without any reference cycles, an object can be simply destroyed once its last
reference is deleted; the finalization and clearing steps are not necessary to
safely reclaim unused objects.  However, it can be useful to automatically call
:c:member:`~PyTypeObject.tp_finalize` and :c:member:`~PyTypeObject.tp_clear`
before destruction anyway because type design is simplified when all objects
always experience the same series of events regardless of whether they
participated in a cyclic isolate.  :term:`CPython` currently only calls
:c:member:`~PyTypeObject.tp_finalize` and :c:member:`~PyTypeObject.tp_clear` as
needed to destroy a cyclic isolate; this may change in a future version.

Functions
---------

To allocate and free memory, see :ref:`allocating-objects`.

.. c:function:: void PyObject_CallFinalizer(PyObject *op)

   Finalizes the object as described in :c:member:`~PyTypeObject.tp_finalize`.
   Call this function (or :c:func:`PyObject_CallFinalizerFromDealloc`) instead
   of calling :c:member:`~PyTypeObject.tp_finalize` directly because this

>    function can deduplicate multiple calls to :c:member:`!tp_finalize`.

Reviewer 3: It might be worth noting here that a call will only be deduplicated if the object is GC tracked. (But I would also note that that's an implementation detail.)

.. c:function:: int PyObject_CallFinalizerFromDealloc(PyObject *op)

   Same as :c:func:`PyObject_CallFinalizer` but meant to be called at the
   beginning of the object's destructor (:c:member:`~PyTypeObject.tp_dealloc`).
   The object's reference count must already be 0.  If the object's finalizer
   resurrects the object, this function returns -1; no further destruction
   should happen.  Otherwise, this function returns 0 and destruction can

>    continue normally.

Reviewer 3: Example usage could be nice. Maybe something like this?

```c
static void
my_dealloc(PyObject *self)
{
    if (PyObject_CallFinalizerFromDealloc(self) < 0)
    {
        // Object got resurrected
        return;
    }
    /* Finalizer passed successfully, rest of deallocation */
    Py_TYPE(self)->tp_free(self);
}
```

Reviewer 2: Maybe “see <tp_dealloc> for example usage”?

              svg.setAttribute('aria-describedby',
                               'life-events-graph-description');
              svg.setAttribute('aria-labelledby', 'life-events-graph-title');
          })();
      </script>

.. only:: epub or not (html or latex)

   .. image:: lifecycle.dot.svg
      :align: center
      :class: invert-in-dark-mode
      :alt: Diagram showing events in an object's life.  Explained in detail
            below.

.. only:: latex

   .. image:: lifecycle.dot.pdf
      :align: center
      :class: invert-in-dark-mode
      :alt: Diagram showing events in an object's life.  Explained in detail
            below.

.. container::
   :name: life-events-graph-description

   Explanation:

   * When a new object is constructed by calling its type:

     #. :c:member:`~PyTypeObject.tp_new` is called to create a new object.
     #. :c:member:`~PyTypeObject.tp_alloc` is directly called by
        :c:member:`~PyTypeObject.tp_new` to allocate the memory for the new
        object.
     #. :c:member:`~PyTypeObject.tp_init` initializes the newly created object.
        :c:member:`!tp_init` can be called again to re-initialize an object, if

>         desired.

Reviewer 2: `tp_new` can be skipped:
```suggestion
        desired. The :c:member:`!tp_init` call can also be skipped entirely,
        for example by Python code calling :py:meth:`~object.__new__`.
```
This also needs an arrow in the diagram.

   * After :c:member:`!tp_init` completes, the object is ready to use.

>    * After the last reference to an object is removed:

Reviewer 2: ```suggestion
   * Some time after the last reference to an object is removed:
```

     #. If an object is not marked as *finalized*, it might be finalized by
        marking it as *finalized* and calling its
        :c:member:`~PyTypeObject.tp_finalize` function.  Python currently does
        not finalize an object when the last reference to it is deleted, but

>         this may change in the future.

Reviewer 3: You should add something like "(unless `PyObject_CallFinalizerFromDealloc` is used in the destructor)"

     #. If the object is marked as finalized,
        :c:member:`~PyTypeObject.tp_clear` might be called to clear references
        held by the object.  Python currently does not clear an object in
        response to the deletion of the last reference, but this may change in
        the future.

>      #. :c:member:`~PyTypeObject.tp_dealloc` is called to destroy the object.

Reviewer 3: Maybe we want to say that it's good practice to just have `tp_dealloc` call into the `tp_clear` slot?

     #. When :c:member:`~PyTypeObject.tp_dealloc` finishes object destruction,
        it directly calls :c:member:`~PyTypeObject.tp_free` (usually set to
        :c:func:`PyObject_Free` or :c:func:`PyObject_GC_Del` automatically as
        appropriate for the type) to deallocate the memory.

   * The :c:member:`~PyTypeObject.tp_finalize` function is permitted to add a
     reference to the object if desired.  If it does, the object is

>      *resurrected*, preventing its pending destruction.  (Only

Reviewer 3: Why is "resurrected" italicized?

     :c:member:`!tp_finalize` is allowed to resurrect an object;
     :c:member:`~PyTypeObject.tp_clear` and

>      :c:member:`~PyTypeObject.tp_dealloc` cannot.)  Resurrecting an object may

Reviewer 3: Well, `tp_dealloc` can through `PyObject_CallFinalizerFromDealloc`.

>      :c:member:`~PyTypeObject.tp_dealloc` cannot.)  Resurrecting an object may

Reviewer 2: This would make the text contradictory: if `tp_finalize` may resurrect an object but `tp_dealloc` may not, then `tp_dealloc` may call `PyObject_CallFinalizerFromDealloc` (which calls `tp_finalize`).

AFAIK, *any* of these may resurrect the object.

Reviewer 3: I don't think `tp_clear` can resurrect, can it? `tp_clear` and `tp_dealloc` should be resurrecting via the finalizer, but if `tp_clear` is called, then `tp_finalize` should have been called by the GC anyway.

     or may not cause the object's *finalized* mark to be removed.  Currently,
     Python does not remove the *finalized* mark from a resurrected object if
     it supports garbage collection (i.e., the :c:macro:`Py_TPFLAGS_HAVE_GC`
     flag is set) but does remove the mark if the object does not support
     garbage collection; either or both of these behaviors may change in the
     future.
   * :c:member:`~PyTypeObject.tp_dealloc` can optionally call
     :c:member:`~PyTypeObject.tp_finalize` via
     :c:func:`PyObject_CallFinalizerFromDealloc` if it wishes to reuse that
     code to help with object destruction.  This is recommended because it
     guarantees that :c:member:`!tp_finalize` is always called before
     destruction.  See the :c:member:`~PyTypeObject.tp_dealloc` documentation

>      for example code.

Reviewer 2: `tp_clear` documentation also suggests that `tp_dealloc` can call `tp_clear`, to avoid code duplication.

Reviewer 2: AFAIK:

For *heap* types (ones defined in Python code, and ones defined via `PyType_From*` functions), `tp_dealloc` defaults to a function called `subtype_dealloc`, which does the whole dance: calls `PyObject_CallFinalizerFromDealloc` and `PyObject_GC_UnTrack`, `__dict__` and delegates to a supertype's dealloc.

For these types, we should, strongly recommend to not implement your own `tp_dealloc`.

`tp_dealloc` is what's set when a user defines  `__del__` in Python code. That's why it's important to call it *once*, and *always*: users expect their custom code to run!

However, it's *not* that important to call in on types you control, where you know users can't provide `__del__` -- that is, on *immutable* types (`Py_TPFLAGS_IMMUTABLETYPE`).

Note that all static (i.e. non-heap) types are immutable.

So, the recommendation to call `PyObject_CallFinalizerFromDealloc` is only applicable when you're defining a heap type with custom `tp_dealloc` -- which should not be recommended.

If you're looking to write `tp_finalize` as a C function, a better recommendation would, IMO, be to instead put the clearing code in `tp_clear`, and call *that* from `tp_dealloc` (making sure that your `tp_clear` is idempotent). I see this PR already makes that recommendation.

   * If the object is a member of a :term:`cyclic isolate` and either
     :c:member:`~PyTypeObject.tp_clear` fails to break the reference cycle or
     the cyclic isolate is not detected (perhaps :func:`gc.disable` was called,
     or the :c:macro:`Py_TPFLAGS_HAVE_GC` flag was erroneously omitted in one
     of the involved types), the objects remain indefinitely uncollectable
     (they "leak").  See :data:`gc.garbage`.

   If the object is marked as supporting garbage collection (the
   :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
   :c:member:`~PyTypeObject.tp_flags`), the following events are also possible:

   * The garbage collector occasionally calls
     :c:member:`~PyTypeObject.tp_traverse` to identify :term:`cyclic isolates
     <cyclic isolate>`.
   * When the garbage collector discovers a :term:`cyclic isolate`, it

>      finalizes one of the objects in the group by marking it as *finalized* and

Reviewer 3: I feel like I've already asked this, but why is this italicized?

     This repeats until the cyclic isolate doesn't exist or all of the objects
     have been finalized.
   * :c:member:`~PyTypeObject.tp_finalize` is permitted to resurrect the object
     by adding a reference from outside the :term:`cyclic isolate`.  The new
     reference causes the group of objects to no longer form a cyclic isolate
     (the reference cycle may still exist, but if it does the objects are no
     longer isolated).
   * When the garbage collector discovers a :term:`cyclic isolate` and all of
     the objects in the group have already been marked as *finalized*, the
     garbage collector clears one or more of the uncleared objects in the group
     (possibly concurrently, but with the :term:`GIL` held) by calling each's
     :c:member:`~PyTypeObject.tp_clear` function.  This repeats as long as the
     cyclic isolate still exists and not all of the objects have been cleared.

Cyclic Isolate Destruction
--------------------------

Listed below are the stages of life of a hypothetical :term:`cyclic isolate`
that continues to exist after each member object is finalized or cleared.  It

> is a bug if a cyclic isolate progresses through all of these stages; it should

Reviewer 3: "memory leak" instead of "bug" is more informative:

```suggestion
is a memory leak if a cyclic isolate progresses through all of these stages; it should
```

   :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
   :c:member:`~PyTypeObject.tp_flags`), the following events are also possible:

   * The garbage collector occasionally calls
     :c:member:`~PyTypeObject.tp_traverse` to identify :term:`cyclic isolates
     <cyclic isolate>`.
   * When the garbage collector discovers a :term:`cyclic isolate`, it
     finalizes one of the objects in the group by marking it as *finalized* and
     calling its :c:member:`~PyTypeObject.tp_finalize` function, if it has one.
     This repeats until the cyclic isolate doesn't exist or all of the objects
     have been finalized.
   * :c:member:`~PyTypeObject.tp_finalize` is permitted to resurrect the object
     by adding a reference from outside the :term:`cyclic isolate`.  The new
     reference causes the group of objects to no longer form a cyclic isolate
     (the reference cycle may still exist, but if it does the objects are no
     longer isolated).
   * When the garbage collector discovers a :term:`cyclic isolate` and all of
     the objects in the group have already been marked as *finalized*, the
     garbage collector clears one or more of the uncleared objects in the group

>      (possibly concurrently, but with the :term:`GIL` held) by calling each's

Reviewer 2: GIL! What GIL?
```suggestion
     (possibly concurrently) by calling each's
```

Reviewer 3: Ha, good catch! I think this predated the "attached thread state" change.

   of the participating objects have failed to break the reference cycle as
   required.

If :c:member:`~PyTypeObject.tp_clear` did not exist, then Python would have no
way to safely break a reference cycle.  Simply destroying an object in a cyclic
isolate would result in a dangling pointer, triggering undefined behavior when
an object referencing the destroyed object is itself destroyed.  The clearing
step makes object destruction a two-phase process: first
:c:member:`~PyTypeObject.tp_clear` is called to partially destroy the objects
enough to detangle them from each other, then
:c:member:`~PyTypeObject.tp_dealloc` is called to complete the destruction.

Unlike clearing, finalization is not a phase of destruction.  A finalized
object must still behave properly by continuing to fulfill its design
contracts.  An object's finalizer is allowed to execute arbitrary Python code,
and is even allowed to prevent the impending destruction by adding a reference.
The finalizer is only contemporaneously related to destruction---it runs just
before destruction, which starts with :c:member:`~PyTypeObject.tp_clear` (if

> called) and concludes with :c:member:`~PyTypeObject.tp_dealloc`.

Reviewer 2: Sadly, “contemporaneously" would not win the “longest word in the docs” contest (there's 
“internationalizing”, “indifferentiability”, “internationalization” and of course “supercalifragilisticexpialidocious”). Let's get it out.
```suggestion
Unlike clearing, finalization is not a phase of destruction.  A finalized
object must still behave properly by continuing to fulfill its design
contracts.  An object's finalizer is allowed to execute arbitrary Python code,
and is even allowed to prevent the impending destruction by adding a reference.
The finalizer is only related to destruction by call order---if it runs, it runs
before destruction, which starts with :c:member:`~PyTypeObject.tp_clear` (if
called) and concludes with :c:member:`~PyTypeObject.tp_dealloc`.
```

.. highlight:: c

.. _life-cycle:

Object Life Cycle
=================

Stages
------

The following is an illustration of the stages of life of an object.  Arrows
indicate a "happens before" relationship.  Octagons indicate functions specific

> to :ref:`garbage collection support <supporting-cycle-detection>`.

Reviewer 4: This sentence seems unnecessary:
```suggestion
The following is an illustration of the stages of life of an object.
Octagons indicate functions specific
to :ref:`garbage collection support <supporting-cycle-detection>`.
```

:c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
:c:member:`~PyTypeObject.tp_flags`), the following stages are also possible:

* The garbage collector occasionally calls
  :c:member:`~PyTypeObject.tp_traverse` to identify :term:`cyclic isolates
  <cyclic isolate>`.
* When the garbage collector discovers a cyclic isolate, it finalizes one of
  the objects in the group by calling its :c:member:`~PyTypeObject.tp_finalize`
  function.  This repeats until the cyclic isolate doesn't exist or all of the
  objects have been finalized.
* The :c:member:`~PyTypeObject.tp_finalize` function can optionally increase
  the object's reference count, causing it (and other objects it references) to
  become resurrected and no longer a member of a cyclic isolate.
* When the garbage collector discovers a cyclic isolate and all of the objects
  in the group have already been finalized, the garbage collector clears one of
  the objects in the group by calling its :c:member:`~PyTypeObject.tp_clear`
  function.  This repeats until the cyclic isolate doesn't exist or all of the
  objects have been cleared.

Functions
---------

To allocate and free memory, see :ref:`Allocating Objects on the Heap

> <allocating-objects>`.

Reviewer 4: The `:ref:` should display the title of the section it refers to, so if the title is already "Allocating Objects on the Heap", you should be able to just do:
```suggestion
To allocate and free memory, see :ref:`allocating-objects`.
```
